After a fresh git clone, running unit tests has 9 failing tests (out of 63)

```
git clone https://github.com/auth0/node-jwks-rsa.git
cd node-jwks-rsa
npm i
npm t

...

  54 passing (235ms)
  9 failing

1) passportJwtSecret
       should not provide a key if token is invalid:
     Uncaught Error [ERR_STREAM_WRITE_AFTER_END] [ERR_STREAM_WRITE_AFTER_END]: write after end
      at writeAfterEnd (_http_outgoing.js:643:15)
      at ServerResponse.end (_http_outgoing.js:763:7)
      at allFailed (node_modules/passport/lib/middleware/authenticate.js:174:11)
      at attempt (node_modules/passport/lib/middleware/authenticate.js:180:28)
      at JwtStrategy.strategy.fail (node_modules/passport/lib/middleware/authenticate.js:302:9)
      at /Users/ottokrus/git/node-jwks-rsa/node_modules/passport-jwt/lib/strategy.js:106:33
      at Object.module.exports [as verify] (node_modules/jsonwebtoken/verify.js:63:12)
      at Function.module.exports [as JwtVerifier] (node_modules/passport-jwt/lib/verify_jwt.js:4:16)
      at /Users/ottokrus/git/node-jwks-rsa/node_modules/passport-jwt/lib/strategy.js:104:25
      at JwtStrategy.secretProvider [as _secretOrKeyProvider] (src/integrations/passport.js:30:14)
      at JwtStrategy.authenticate (node_modules/passport-jwt/lib/strategy.js:99:10)
      at attempt (node_modules/passport/lib/middleware/authenticate.js:366:16)
      at authenticate (node_modules/passport/lib/middleware/authenticate.js:367:7)
      at Layer.handle [as handle_request] (node_modules/express/lib/router/layer.js:95:5)
      at next (node_modules/express/lib/router/route.js:137:13)
      at app.get.passport.authenticate.session (tests/passport.tests.js:60:9)
      at Layer.handle [as handle_request] (node_modules/express/lib/router/layer.js:95:5)
      at next (node_modules/express/lib/router/route.js:137:13)
      at Route.dispatch (node_modules/express/lib/router/route.js:112:3)
      at Layer.handle [as handle_request] (node_modules/express/lib/router/layer.js:95:5)
      at /Users/ottokrus/git/node-jwks-rsa/node_modules/express/lib/router/index.js:281:22
      at Function.process_params (node_modules/express/lib/router/index.js:335:12)
      at next (node_modules/express/lib/router/index.js:275:10)
      at expressInit (node_modules/express/lib/middleware/init.js:40:5)
      at Layer.handle [as handle_request] (node_modules/express/lib/router/layer.js:95:5)
      at trim_prefix (node_modules/express/lib/router/index.js:317:13)
      at /Users/ottokrus/git/node-jwks-rsa/node_modules/express/lib/router/index.js:284:7
      at Function.process_params (node_modules/express/lib/router/index.js:335:12)
      at next (node_modules/express/lib/router/index.js:275:10)
      at query (node_modules/express/lib/middleware/query.js:45:5)
      at Layer.handle [as handle_request] (node_modules/express/lib/router/layer.js:95:5)
      at trim_prefix (node_modules/express/lib/router/index.js:317:13)
      at /Users/ottokrus/git/node-jwks-rsa/node_modules/express/lib/router/index.js:284:7
      at Function.process_params (node_modules/express/lib/router/index.js:335:12)
      at next (node_modules/express/lib/router/index.js:275:10)
      at Function.handle (node_modules/express/lib/router/index.js:174:3)
      at Function.handle (node_modules/express/lib/application.js:174:10)
      at Server.app (node_modules/express/lib/express.js:39:9)
      at parserOnIncoming (_http_server.js:779:12)
      at HTTPParser.parserOnHeadersComplete (_http_common.js:117:17)

  2) passportJwtSecret
       should not provide a key if JWKS endpoint returned multiple keys and no KID was provided:
     Uncaught Error [ERR_STREAM_WRITE_AFTER_END] [ERR_STREAM_WRITE_AFTER_END]: write after end
      at writeAfterEnd (_http_outgoing.js:643:15)
      at ServerResponse.end (_http_outgoing.js:763:7)
      at allFailed (node_modules/passport/lib/middleware/authenticate.js:174:11)
      at attempt (node_modules/passport/lib/middleware/authenticate.js:180:28)
      at JwtStrategy.strategy.fail (node_modules/passport/lib/middleware/authenticate.js:302:9)
      at /Users/ottokrus/git/node-jwks-rsa/node_modules/passport-jwt/lib/strategy.js:106:33
      at /Users/ottokrus/git/node-jwks-rsa/node_modules/jsonwebtoken/verify.js:106:14
      at getSecret (node_modules/jsonwebtoken/verify.js:90:14)
      at Object.module.exports [as verify] (node_modules/jsonwebtoken/verify.js:94:10)
      at Function.module.exports [as JwtVerifier] (node_modules/passport-jwt/lib/verify_jwt.js:4:16)
      at /Users/ottokrus/git/node-jwks-rsa/node_modules/passport-jwt/lib/strategy.js:104:25
      at /Users/ottokrus/git/node-jwks-rsa/src/integrations/passport.js:35:43
      at handleSigningKeyError (src/integrations/passport.js:8:12)
      at Immediate._onImmediate (src/integrations/passport.js:35:16)
      at processImmediate (internal/timers.js:458:21)

  3) passportJwtSecret
       should allow returning an error if key not found:

      Uncaught AssertionError: expected 'OK' to equal 'this is bad'
      + expected - actual

      -OK
      +this is bad
      
      at Test.<anonymous> (tests/passport.tests.js:286:29)
      at Test.assert (node_modules/supertest/lib/test.js:181:6)
      at localAssert (node_modules/supertest/lib/test.js:131:12)
      at /Users/ottokrus/git/node-jwks-rsa/node_modules/supertest/lib/test.js:128:5
      at Test.Request.callback (node_modules/superagent/lib/node/index.js:716:12)
      at IncomingMessage.<anonymous> (node_modules/superagent/lib/node/index.js:916:18)
      at endReadableNT (_stream_readable.js:1225:12)
      at processTicksAndRejections (internal/process/task_queues.js:84:21)

  4) passportJwtSecret
       should work if the token matches a signing key:
     Uncaught AssertionError: expected undefined to equal 'john'
      at Test.<anonymous> (tests/passport.tests.js:322:33)
      at Test.assert (node_modules/supertest/lib/test.js:181:6)
      at localAssert (node_modules/supertest/lib/test.js:131:12)
      at /Users/ottokrus/git/node-jwks-rsa/node_modules/supertest/lib/test.js:128:5
      at Test.Request.callback (node_modules/superagent/lib/node/index.js:728:3)
      at IncomingMessage.<anonymous> (node_modules/superagent/lib/node/index.js:916:18)
      at endReadableNT (_stream_readable.js:1225:12)
      at processTicksAndRejections (internal/process/task_queues.js:84:21)

  5) passportJwtSecret
       should work if the JWKS endpoint returns a single key and no KID is provided:
     Uncaught AssertionError: expected undefined to equal 'john'
      at Test.<anonymous> (tests/passport.tests.js:358:33)
      at Test.assert (node_modules/supertest/lib/test.js:181:6)
      at localAssert (node_modules/supertest/lib/test.js:131:12)
      at /Users/ottokrus/git/node-jwks-rsa/node_modules/supertest/lib/test.js:128:5
      at Test.Request.callback (node_modules/superagent/lib/node/index.js:728:3)
      at IncomingMessage.<anonymous> (node_modules/superagent/lib/node/index.js:916:18)
      at endReadableNT (_stream_readable.js:1225:12)
      at processTicksAndRejections (internal/process/task_queues.js:84:21)

  6) koaJwtSecret
       should not provide a key if JWKS endpoint returned multiple keys and no KID was provided:

      Uncaught AssertionError: expected 'invalid signature' to equal 'No KID specified and JWKS endpoint returned more than 1 key'
      + expected - actual

      -invalid signature
      +No KID specified and JWKS endpoint returned more than 1 key
      
      at Test.<anonymous> (tests/koa.tests.js:93:29)
      at Test.assert (node_modules/supertest/lib/test.js:181:6)
      at localAssert (node_modules/supertest/lib/test.js:131:12)
      at /Users/ottokrus/git/node-jwks-rsa/node_modules/supertest/lib/test.js:128:5
      at Test.Request.callback (node_modules/superagent/lib/node/index.js:728:3)
      at IncomingMessage.<anonymous> (node_modules/superagent/lib/node/index.js:916:18)
      at endReadableNT (_stream_readable.js:1225:12)
      at processTicksAndRejections (internal/process/task_queues.js:84:21)

  7) koaJwtSecret
       should not authenticate the user if KID matches but the keys don't:

      Uncaught AssertionError: expected 'Unable to find a signing key that matches \'123\'' to equal 'invalid signature'
      + expected - actual

      -Unable to find a signing key that matches '123'
      +invalid signature
      
      at Test.<anonymous> (tests/koa.tests.js:137:29)
      at Test.assert (node_modules/supertest/lib/test.js:181:6)
      at localAssert (node_modules/supertest/lib/test.js:131:12)
      at /Users/ottokrus/git/node-jwks-rsa/node_modules/supertest/lib/test.js:128:5
      at Test.Request.callback (node_modules/superagent/lib/node/index.js:728:3)
      at IncomingMessage.<anonymous> (node_modules/superagent/lib/node/index.js:916:18)
      at endReadableNT (_stream_readable.js:1225:12)
      at processTicksAndRejections (internal/process/task_queues.js:84:21)

  8) koaJwtSecret
       should allow returning an error if key not found:

      Uncaught AssertionError: expected 'Not Found' to equal 'this is bad'
      + expected - actual

      -Not Found
      +this is bad
      
      at Test.<anonymous> (tests/koa.tests.js:167:29)
      at Test.assert (node_modules/supertest/lib/test.js:181:6)
      at localAssert (node_modules/supertest/lib/test.js:131:12)
      at /Users/ottokrus/git/node-jwks-rsa/node_modules/supertest/lib/test.js:128:5
      at Test.Request.callback (node_modules/superagent/lib/node/index.js:728:3)
      at IncomingMessage.<anonymous> (node_modules/superagent/lib/node/index.js:916:18)
      at endReadableNT (_stream_readable.js:1225:12)
      at processTicksAndRejections (internal/process/task_queues.js:84:21)

  9) koaJwtSecret
       should work if the JWKS endpoint returns a single key and no KID is provided:
     Uncaught AssertionError: expected undefined to equal 'john'
      at Test.<anonymous> (tests/koa.tests.js:220:33)
      at Test.assert (node_modules/supertest/lib/test.js:181:6)
      at localAssert (node_modules/supertest/lib/test.js:131:12)
      at /Users/ottokrus/git/node-jwks-rsa/node_modules/supertest/lib/test.js:128:5
      at Test.Request.callback (node_modules/superagent/lib/node/index.js:728:3)
      at IncomingMessage.<anonymous> (node_modules/superagent/lib/node/index.js:916:18)
      at endReadableNT (_stream_readable.js:1225:12)
      at processTicksAndRejections (internal/process/task_queues.js:84:21)
```

I just cloned and ran the tests on a fresh repo using node v12 and received two failing tests out of 68.
node version:  v12.19.0

Maybe some updates are need to how async tests are handled in mocha? 

```


  66 passing (6s)
  2 failing

  1) JwksClient (cache)
       #getSigningKey
         should cache requests per kid
           should ignore the cache when the KID isnt cached and make a requst:
     Uncaught AssertionError: expected undefined to equal 'ENOTFOUND'
      at Immediate.<anonymous> (tests/cache.tests.js:41:31)
      at processImmediate (internal/timers.js:463:21)

  2) JwksClient
       #getKeys
         when using a proxy
           should fail when proxy address is not reachable:
     Error: Timeout of 5000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves. (/Users/michael/git/scratch/node-jwks-rsa/tests/jwksClient.tests.js)
      at listOnTimeout (internal/timers.js:554:17)
      at processTimers (internal/timers.js:497:7)
```

--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -1,17 +1,26 @@
-version: 2
+version: 2.1
+
 workflows:
-  matrix-tests:
+  version: 2
+  test:
     jobs:
-      - node/test:
-          version: 14.15.4
-      - node/test:
-          version: 12.20.1
-      - node/test:
-          version: 10.13.0
+      - test:
+          name: test_node10
+          v: "10"
+      - test:
+          name: test_node12
+          v: "12"
+      - test:
+          name: test_node14
+          v: "14"
 jobs:
-  build:
+  test:
+    parameters:
+      v:
+        type: string
+        default: "12"
     docker:
-      - image: circleci/node:10
+      - image: circleci/node:<< parameters.v >>
     environment:
       LANG: en_US.UTF-8
     steps:
@@ -19,7 +28,7 @@ jobs:
       - run:
           name: Update NPM version
           command: sudo npm install -g npm@latest
-      - restore-cache:
+      - restore_cache:
           name: Restore NPM Package Cache
           keys:
             - npm-packages-{{ checksum "package-lock.json" }}
@@ -27,7 +36,7 @@ jobs:
       - run:
           name: Install Dependencies
           command: npm install
-      - save-cache:
+      - save_cache:
           name: Save NPM Package Cache
           key: npm-packages-{{ checksum "package-lock.json" }}
           paths:

--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -1,4 +1,13 @@
 version: 2
+workflows:
+  matrix-tests:
+    jobs:
+      - node/test:
+          version: 14.15.4
+      - node/test:
+          version: 12.20.1
+      - node/test:
+          version: 10.13.0
 jobs:
   build:
     docker:

--- a/package.json
+++ b/package.json
@@ -4,6 +4,9 @@
   "description": "Library to retrieve RSA public keys from a JWKS endpoint",
   "main": "lib/index.js",
   "types": "index.d.ts",
+  "engines": {
+    "node": ">=10 < 13 || >=14"
+  },
   "dependencies": {
     "@types/express-jwt": "0.0.42",
     "axios": "^0.21.1",

--- a/tests/passport.tests.js
+++ b/tests/passport.tests.js
@@ -63,11 +63,12 @@ describe('passportJwtSecret', () => {
       )
     );
 
+    let expectedFlashMessage;
     app.get(
       '/',
       (req, res, next) => {
         req.flash = (type, msg) => {
-          res.send(msg);
+          expectedFlashMessage = msg;
         };
         next();
       },
@@ -81,8 +82,8 @@ describe('passportJwtSecret', () => {
       .get('/')
       .set('Authorization', 'Bearer abc')
       .expect(401)
-      .end((err, res) => {
-        expect(res.text).to.equal('jwt malformed');
+      .end(() => {
+        expect(expectedFlashMessage).to.equal('jwt malformed');
         done();
       });
   });
@@ -101,11 +102,12 @@ describe('passportJwtSecret', () => {
       )
     );
 
+    let expectedFlashMessage;
     app.get(
       '/',
       (req, res, next) => {
         req.flash = (type, msg) => {
-          res.send(msg);
+          expectedFlashMessage = msg;
         };
         next();
       },
@@ -121,8 +123,8 @@ describe('passportJwtSecret', () => {
       .get('/')
       .set('Authorization', `Bearer ${token}`)
       .expect(401)
-      .end((err, res) => {
-        expect(res.text).to.equal('secret or public key must be provided');
+      .end(() => {
+        expect(expectedFlashMessage).to.equal('secret or public key must be provided');
         done();
       });
   });
@@ -141,12 +143,13 @@ describe('passportJwtSecret', () => {
       )
     );
 
+    let expectedFlashMessage;
     app.get(
       '/',
       (req, res, next) => {
         req.flash = (type, msg) => {
-          res.send(msg);
-        };
+          expectedFlashMessage = msg;
+        },
         next();
       },
       passport.authenticate('jwt', { session: false, failureFlash: true }),
@@ -162,9 +165,9 @@ describe('passportJwtSecret', () => {
       .get('/')
       .set('Authorization', `Bearer ${token}`)
       .expect(401)
-      .end((err, res) => {
-        expect(res.text).to.equal('secret or public key must be provided');
-        done();
+      .end((err) => {
+        expect(expectedFlashMessage).to.equal('secret or public key must be provided');
+        done(err);
       });
   });
 
@@ -184,11 +187,12 @@ describe('passportJwtSecret', () => {
       )
     );
 
+    let expectedFlashMessage;
     app.get(
       '/',
       (req, res, next) => {
         req.flash = (type, msg) => {
-          res.send(msg);
+          expectedFlashMessage = msg;
         };
         next();
       },
@@ -205,8 +209,8 @@ describe('passportJwtSecret', () => {
       .get('/')
       .set('Authorization', `Bearer ${token}`)
       .expect(401)
-      .end((err, res) => {
-        expect(res.text).to.equal('secret or public key must be provided');
+      .end(() => {
+        expect(expectedFlashMessage).to.equal('secret or public key must be provided');
         done();
       });
   });
@@ -227,11 +231,12 @@ describe('passportJwtSecret', () => {
       )
     );
 
+    let expectedFlashMessage;
     app.get(
       '/',
       (req, res, next) => {
         req.flash = (type, msg) => {
-          res.send(msg);
+          expectedFlashMessage = msg;
         };
         next();
       },
@@ -248,8 +253,8 @@ describe('passportJwtSecret', () => {
       .get('/')
       .set('Authorization', `Bearer ${token}`)
       .expect(401)
-      .end((err, res) => {
-        expect(res.text).to.equal('invalid signature');
+      .end(() => {
+        expect(expectedFlashMessage).to.equal('invalid signature');
         done();
       });
   });
@@ -272,12 +277,13 @@ describe('passportJwtSecret', () => {
         (jwt_payload, done) => done(null, jwt_payload)
       )
     );
-
+    
+    let expectedFlashMessage;
     app.get(
       '/',
       (req, res, next) => {
         req.flash = (type, msg) => {
-          res.send(msg);
+          expectedFlashMessage = msg;
         };
         next();
       },
@@ -294,8 +300,8 @@ describe('passportJwtSecret', () => {
       .get('/')
       .set('Authorization', `Bearer ${token}`)
       .expect(401)
-      .end((err, res) => {
-        expect(res.text).to.equal('this is bad');
+      .end(() => {
+        expect(expectedFlashMessage).to.equal('this is bad');
         done();
       });
   });

